libnmstate/nm/sriov.py

```diff
@@ -43,11 +43,79 @@ def create_setting(iface_state, base_con_profile):
         if not sriov_setting:
             sriov_setting = nmclient.NM.SettingSriov.new()
 
+        vfs_config = sriov_config.get(Ethernet.SRIOV.VFS_SUBTREE, [])
+        _create_sriov_vfs_in_setting(vfs_config, sriov_setting)
+        _remove_sriov_vfs_in_setting(vfs_config, sriov_setting)
+
         sriov_setting.props.total_vfs = sriov_config[Ethernet.SRIOV.TOTAL_VFS]
 
     return sriov_setting
 
 
+def _create_sriov_vfs_in_setting(vfs_config, sriov_setting):
+    vf_object_ids = [vf.get_index() for vf in sriov_setting.props.vfs]
+
+    for vf_config in vfs_config:
+        vf_already_exists = False
+        vf_id = vf_config[Ethernet.SRIOV.VFS.ID]
+        if vf_id in vf_object_ids:
+            vf_already_exists = True
+
+        vf_object = nmclient.NM.SriovVF.new(vf_config[Ethernet.SRIOV.VFS.ID])
+
+        if vf_config.get(Ethernet.SRIOV.VFS.MAC, None) is not None:
+            vf_object.set_attribute(
+                nmclient.NM.SRIOV_VF_ATTRIBUTE_MAC,
+                nmclient.GLib.Variant.new_string(
+                    vf_config[Ethernet.SRIOV.VFS.MAC]
+                ),
+            )
+        if vf_config.get(Ethernet.SRIOV.VFS.TRUST, None) is not None:
+            vf_object.set_attribute(
+                nmclient.NM.SRIOV_VF_ATTRIBUTE_TRUST,
+                nmclient.GLib.Variant.new_boolean(
+                    vf_config[Ethernet.SRIOV.VFS.TRUST]
+                ),
+            )
+        if vf_config.get(Ethernet.SRIOV.VFS.MAX_TX_RATE, None) is not None:
+            vf_object.set_attribute(
+                nmclient.NM.SRIOV_VF_ATTRIBUTE_MAX_TX_RATE,
+                nmclient.GLib.Variant.new_uint32(
+                    vf_config[Ethernet.SRIOV.VFS.MAX_TX_RATE]
+                ),
+            )
+        if vf_config.get(Ethernet.SRIOV.VFS.MIN_TX_RATE, None) is not None:
+            vf_object.set_attribute(
+                nmclient.NM.SRIOV_VF_ATTRIBUTE_MIN_TX_RATE,
+                nmclient.GLib.Variant.new_uint32(
+                    vf_config[Ethernet.SRIOV.VFS.MIN_TX_RATE]
+                ),
+            )
+        if vf_config.get(Ethernet.SRIOV.VFS.SPOOF_CHECK, None) is not None:
+            vf_object.set_attribute(
+                nmclient.NM.SRIOV_VF_ATTRIBUTE_SPOOF_CHECK,
+                nmclient.GLib.Variant.new_boolean(
+                    vf_config[Ethernet.SRIOV.VFS.SPOOF_CHECK]
+                ),
+            )
+
+        if vf_already_exists:
+            sriov_setting.remove_vf_by_index(vf_id)
+
+        sriov_setting.add_vf(vf_object)
+
+
+def _remove_sriov_vfs_in_setting(vfs_config, sriov_setting):
+    vf_object_ids = {vf.get_index() for vf in sriov_setting.props.vfs}
+    vf_config_ids = {
+        vf_config[Ethernet.SRIOV.VFS.ID] for vf_config in vfs_config
+    }
+    vf_ids_to_remove = vf_object_ids.difference(vf_config_ids)
```

Comment:
This is equivalent to `vf_object_ids - vf_config_ids`

Reply:
Done!